File: app/src/main/java/io/github/ovso/righttoknow/ui/main/violationfacility/ViolationFacilityAdapter.java
package io.github.ovso.righttoknow.ui.main.violationfacility;

import android.view.View;
import android.widget.TextView;
import butterknife.BindView;
import io.github.ovso.righttoknow.R;
import io.github.ovso.righttoknow.framework.adapter.BaseAdapterDataModel;
import io.github.ovso.righttoknow.framework.adapter.BaseAdapterView;
import io.github.ovso.righttoknow.framework.adapter.BaseRecyclerAdapter;
import io.github.ovso.righttoknow.framework.adapter.OnRecyclerItemClickListener;
import io.github.ovso.righttoknow.ui.main.violationfacility.model.VioFac;
import java.util.ArrayList;
import java.util.List;
import lombok.Setter;

public class ViolationFacilityAdapter extends BaseRecyclerAdapter
    implements BaseAdapterView, BaseAdapterDataModel<VioFac> {

  private List<VioFac> items = new ArrayList<>();

  @Override protected BaseViewHolder createViewHolder(View view, int viewType) {
    return new ViolationFacilityViewHolder(view);
  }

  @Override public int getLayoutRes(int viewType) {
    return R.layout.fragment_violation_item;
  }

  @Override public void onBindViewHolder(BaseViewHolder baseHolder, int position) {
    if (baseHolder instanceof ViolationFacilityViewHolder) {
      ViolationFacilityViewHolder holder = (ViolationFacilityViewHolder) baseHolder;
      final VioFac fac = items.get(position);
      holder.orderTextView.setText(String.valueOf(fac.getOrder()));
      holder.sidoTextView.setText(fac.getSido());
      holder.sigunguTextView.setText(fac.getSigungu());
      holder.facNameTextView.setText(fac.getFac_name());
      holder.typeTextView.setText(fac.getType());
      holder.itemView.setOnClickListener(view -> onRecyclerItemClickListener.onItemClick(fac));
    }
  }

  @Override public int getItemCount() {
    return getSize();
  }

  @Override public void refresh() {
    notifyDataSetChanged();
  }

  @Override public void add(VioFac item) {
    items.add(item);
  }

  @Override public void addAll(List<VioFac> items) {
    this.items.addAll(items);
  }

  @Override public VioFac remove(int position) {
    return items.remove(position);
  }

  @Override public VioFac getItem(int position) {
    return items.get(position);
  }

  @Override public void add(int index, VioFac item) {
    items.add(index, item);
  }

  @Override public int getSize() {
    return items.size();
  }

  @Override public void clear() {
    items.clear();
  }

  @Setter private OnRecyclerItemClickListener<VioFac> onRecyclerItemClickListener;

  final static class ViolationFacilityViewHolder extends BaseViewHolder {
    @BindView(R.id.order_textview) TextView orderTextView;
    @BindView(R.id.sido_textview) TextView sidoTextView;
    @BindView(R.id.sigungu_textview) TextView sigunguTextView;
    @BindView(R.id.fac_name_textview) TextView facNameTextView;
    @BindView(R.id.type_textview) TextView typeTextView;

    public ViolationFacilityViewHolder(View itemView) {
      super(itemView);
    }
  }
}

File: app/src/main/java/io/github/ovso/righttoknow/ui/main/violationfacility/ViolationFacilityAdapter.kt
package io.github.ovso.righttoknow.ui.main.violationfacility

import android.view.View
import io.github.ovso.righttoknow.R
import io.github.ovso.righttoknow.framework.adapter.BaseAdapterDataModel
import io.github.ovso.righttoknow.framework.adapter.BaseAdapterView
import io.github.ovso.righttoknow.framework.adapter.BaseRecyclerAdapter
import io.github.ovso.righttoknow.framework.adapter.OnRecyclerItemClickListener
import io.github.ovso.righttoknow.ui.main.violationfacility.model.VioFac
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.fragment_violation_item.*
import java.util.*

class ViolationFacilityAdapter : BaseRecyclerAdapter(), BaseAdapterView,
  BaseAdapterDataModel<VioFac> {
  private val items: MutableList<VioFac> = ArrayList()
  override fun createViewHolder(view: View, viewType: Int): BaseViewHolder {
    return ViolationFacilityViewHolder(view)
  }

  override fun getLayoutRes(viewType: Int): Int {
    return R.layout.fragment_violation_item
  }

  override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
    if (holder is ViolationFacilityViewHolder) {
      holder.onBindViewHolder(getItem(position))
      holder.itemView.setOnClickListener {
        onRecyclerItemClickListener!!.onItemClick(
          getItem(
            position
          )
        )
      }
    }
  }

  override fun getItemCount(): Int {
    return size
  }

  override fun refresh() {
    notifyDataSetChanged()
  }

  override fun add(item: VioFac) {
    items.add(item)
  }

  override fun addAll(items: List<VioFac>) {
    this.items.addAll(items)
  }

  override fun remove(position: Int): VioFac {
    return items.removeAt(position)
  }

  override fun getItem(position: Int): VioFac {
    return items[position]
  }

  override fun add(index: Int, item: VioFac) {
    items.add(index, item)
  }

  override val size: Int
    get() = items.size

  override fun clear() {
    items.clear()
  }

  var onRecyclerItemClickListener: OnRecyclerItemClickListener<VioFac>? = null

  internal class ViolationFacilityViewHolder(override val containerView: View?) :
    BaseViewHolder(containerView!!), LayoutContainer {
    fun onBindViewHolder(item: VioFac) {
      order_textview.text = item.order.toString()
      sido_textview.text = item.sido
      sigungu_textview.text = item.sigungu
      fac_name_textview.text = item.fac_name
      type_textview.text = item.type
    }
  }
}
